src/libraries/Common/src/Interop/Windows/Activeds/Interop.ADsGetLastError.cs

```diff
@@ -0,0 +1,13 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Activeds
+    {
+        [LibraryImport(Libraries.Activeds, StringMarshalling = StringMarshalling.Utf16)]
+        internal static unsafe partial int ADsGetLastError(out int error, char* errorBuffer, int errorBufferLength, char* nameBuffer, int nameBufferLength);
```

Comment:
What are the desired semantics here for `nameBuffer`?

Reply:
I don't rightly know - I lifted this code out as-is :\

Reply:
This could be changed a bit, but I don't think it is worth it for right now. Thanks!

Reply:
I'd be curious how if you have the time to explain or point me to a resource that explains it, even if it doesn't happen in this PR, for my own learning - no worries one way or the other!

Reply:
If you look at the use cases they are all passing in a caller allocated buffer, either through `stackalloc` or via an on stack `char`. The entire point of `LibraryImport` is to make interop consistent, safe and the lowest cost possible. This example could improve the safety dimension by accepting a `char[]` and the generated code will pin the array. It is the pattern I would normally recommend, but it does come with some cost (i.e., allocating an array). Another option here is one could leverage the [`Span<T>` marshaller](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.marshalling.spanmarshaller-2.managedtounmanagedin) with `stackalloc` and remove the explicit pointer in the signature and in the caller code.

Reply:
Understood, thanks for taking the time to explain that, and I didn't know `LibraryImport` code gen worked with Span and co. now!

Something like this, then?

```suggestion
        internal static partial int ADsGetLastError(out int error, Span<char> errorBuffer, int errorBufferLength, Span<char> nameBuffer, int nameBufferLength);
```

Then change the call sites to the follow:
```diff
diff --git a/src/libraries/System.DirectoryServices/src/System/DirectoryServices/DirectoryServicesCOMException.cs b/src/libraries/System.DirectoryServices/src/System/DirectoryServices/DirectoryServicesCOMException.cs
index 8e11602eb2a..516d13e6424 100644

@@ -53,14 +53,15 @@ internal static Exception CreateFormattedComException(int hr)
             return CreateFormattedComException(new COMException(errorMsg, hr));
         }

-        internal static unsafe Exception CreateFormattedComException(COMException e)
+        internal static Exception CreateFormattedComException(COMException e)
         {
             // get extended error information
             const int ErrorBufferLength = 256;
-            char* errorBuffer = stackalloc char[ErrorBufferLength];
-            char nameBuffer = '\0';
+            Span<char> errorBuffer = stackalloc char[ErrorBufferLength];
+
+            Span<char> nameBuffer = Span<char>.Empty;
             int error = 0;
-            Interop.Activeds.ADsGetLastError(out error, errorBuffer, ErrorBufferLength, &nameBuffer, 0);
+            Interop.Activeds.ADsGetLastError(out error, errorBuffer, ErrorBufferLength, nameBuffer, 0);

             if (error != 0)
                 return new DirectoryServicesCOMException(new string(errorBuffer), error, e);
```

```diff
diff --git a/src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs b/src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs
index 4b9a4884caa..8067d2e13cf 100644
--- a/src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs
+++ b/src/libraries/System.DirectoryServices/src/System/DirectoryServices/SearchResultCollection.cs
@@ -401,11 +401,12 @@ private void CleanLastError()
                 Interop.Activeds.ADsSetLastError(Interop.Errors.ERROR_SUCCESS, null, null);
             }

-            private unsafe int GetLastError(ref int errorCode)
+            private int GetLastError(ref int errorCode)
             {
-                char c1 = '\0', c2 = '\0';
+                Span<char> c1 = Span<char>.Empty;
+                Span<char> c2 = Span<char>.Empty;
                 errorCode = Interop.Errors.ERROR_SUCCESS;
-                return Interop.Activeds.ADsGetLastError(out errorCode, &c1, 0, &c2, 0);
+                return Interop.Activeds.ADsGetLastError(out errorCode, c1, 0, c2, 0);
             }
         }
     }
```

Reply:
The length can be referenced through the `Span<T>` instance instead of specifying a length in a disjoint manner. For example,

```csharp
c1, c1.Length
```

Reply:
Done in 497bbb6ed478fb2bcdc47da75d885cb4a29b1ca8, thanks again for the guidance.

Reply:
Ah, this isn't quite right - it's causing `COMException: Invalid pointer` for empty spans - is there a way of having them passed as not null when they're empty?

Reply:
I could make the values I pass in `Span<char> arg = Array<char>.Empty;` rather than `Span<char>.Empty` to fix this I think (since "this marshaller will pass a non-null value for a zero-length span if the span was constructed with a non-null value.") , but it seems it would be better if that could be fixed on the interop method definition side?

Reply:
On reading the description of https://github.com/dotnet/runtime/issues/69281 - I think what I actually want here is an `ArrayMarshaller` so the interop method definition should accept `char[]` instead after all?

Reply:
Fixed in 042487f0cef9828dea36df4adcb1c94f0fbd11de and explained my reasoning for the choices I've made in the commit message, please let me know if I'm still missing the mark!

Reply:
@jkoritzinsky @jtschuster Here is another occurrence of the empty `Span<T>` logic conflicting with `null` semantics. Just FYI.